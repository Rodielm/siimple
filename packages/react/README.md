# @siimple/react (Experimental)

Style your React components using siimple.

> **Note**: this is an experimental package. API may change at any time.

## Installation

Install this package using **npm**:

```bash
 npm install --save @siimple/react
```

## Usage

You can use `ThemeProvider` component for using your theme, and the `styled` function to create styled components:

```js
import React from "react";
import {ThemeProvider, styled} from "@siimple/react";

const theme = {
    colors: {
        primary: "cadetblue",
    },
};

const Button = styled("a", {
    backgroundColor: "primary",
    color: "white",
    padding: "1rem",
    width: "100%",
});

const App = () => (
    <ThemeProvider theme={theme}>
        <Button href="/">Hello world</Button>
    </ThemeProvider>
);
```

## API

### ThemeProvider

A simple React component that accepts a **siimple theme object** as a prop and pass it top-down using [Context](https://reactjs.org/docs/context.html).

```jsx
import {ThemeProvider} from "@siimple/react";

export default App = () => (
    <ThemeProvider theme={theme}>
        ...
    </ThemeProvider>
);
```

### styled(type, css)

Generate a styled React component with the specified type and CSS styles. If the returned component is used as a child of `<ThemeProvider>`, the theme provided will be used to generate the styles.

```jsx
import {styled} from "@siimple/react";

const ButtonLink = styled("a", {
    color: "primary",
    display: "block",
    padding: "2rem",
    textDecoration: "none",
    "&:hover": {
        cursor: "pointer",
    },
});

const App = () => (
    <div align="center">
        <ButtonLink href="/login">
            Login
        </ButtonLink>
    </div>
);
```

### useTheme()

A React [Hook](https://reactjs.org/docs/hooks-intro.html) that returns the current theme provided to `<ThemeProvider>`.

```jsx
import {useTheme} from "@siimple/react";

export const Text = props => {
    const theme = useTheme();

    return (
        <span style={{color: theme.colors.primary}}>
            {props.children}
        </span>
    );
};
```

### useCss(css)

A React Hook that accepts a CSS object and returns a classname for the provided styles. This hook will use the theme object provided in the `<ThemeProvider>` component (if used).

```jsx
import React from "react";
import {useCss} from "@siimple/react";

const theme = {
    colors: {
        primary: "cadetblue",
    },
};

const Button = props => {
    const className = useCss({
        backgroundColor: "primary",
        borderRadius: "0.5rem",
        color: "white",
        padding: "2re",
        width: "100%",
    });
    return (
        <a className={className} href={props.href}>
            {props.children}
        </a>
    );
};

const App = () => (
    <ThemeProvider theme={theme}>
        <Button href="/">Hello world</Button>
    </ThemeProvider>
);
```

### classNames(obj)

A tiny utility for conditionally joining class names.

```js
import {classNames} from "@siimple/react";

const names = classNames({
    "foo": true,
    "bar": trueCondition === true,
    "baz": null,
});
// names === "foo bar"
```

## License

[MIT License](https://github.com/jmjuanes/siimple/blob/main/LICENSE).
