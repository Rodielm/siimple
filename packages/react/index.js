import React from "react";
import {classNames, createCache, registerCss} from "@siimple/styled";

// Contexts
export const ThemeContext = React.createContext({});
export const CacheContext = React.createContext(
    createCache({key: "siimple-react"}),
);

// Internal css compiler
const __useCss = (css, isGlobal) => {
    const theme = React.useContext(ThemeContext);
    const cache = React.useContext(CacheContext);

    return registerCss(css, theme, cache, isGlobal, false);
};

// Available hooks
export const useTheme = () => React.useContext(ThemeContext);
export const useCss = css => __useCss(css, false);
export const useGlobalCss = css => __useCss(css, true);

// Theme provider component
export const ThemeProvider = props => {
    let theme = useTheme();
    if (props.theme && typeof props.theme === "object" && props.theme !== theme) {
        theme = props.theme;
    }
    return React.createElement(ThemeContext.Provider, {value: theme}, props.children);
};

// HOC that wraps a component and adds the current theme as a prop
export const withTheme = Component => {
    const name = Component.displayName || Component.name || "Component";
    const ThemedComponent = React.forwardRef((props, ref) => {
        const theme = useTheme();

        return React.createElement(Component, {
            ...props,
            theme: theme,
            ref: ref,
        });
    });
    ThemedComponent.displayName = `withTheme(${name})`;
    return ThemedComponent;
};

// Create an styled element
export const styled = (type = "div", css) => {
    return React.forwardRef((props, ref) => {
        const {css: cssProp, ...rest} = props;
        const className = useCss({
            ...css,
            ...(cssProp || {}),
        });

        return React.createElement(type, {
            ...rest,
            ref: ref,
            className: classNames(props.className, className),
        });
    });
};
