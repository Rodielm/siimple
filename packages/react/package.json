{
    "name": "@siimple/react",
    "version": "0.3.0",
    "description": "Style your React components using siimple",
    "author": "Josemi Juanes <josemi@siimple.xyz>",
    "type": "module",
    "license": "MIT",
    "repository": "https://github.com/jmjuanes/siimple",
    "bugs": "https://github.com/jmjuanes/siimple/issues",
    "homepage": "https://www.siimple.xyz",
    "main": "index.js",
    "keywords": [
        "siimple",
        "react"
    ],
    "dependencies": {
        "@siimple/css": "^0.3.0",
        "react": "^17.0.2"
    }
}
