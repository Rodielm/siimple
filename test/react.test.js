/**
 * @jest-environment jsdom
 */

import renderer from "react-test-renderer";
import {ThemeProvider, useTheme, styled} from "@siimple/react";

const selector = `style[data-siimple="siimple-react"]`;

describe("ThemeProvider", () => {
    it("should render", () => {
        const component = renderer.create((
            <ThemeProvider theme={{}}>
                <div>Hello world</div>
            </ThemeProvider>
        ));

        expect(component.toJSON()).toMatchSnapshot();
    });
});

describe("useTheme", () => {
    it("should return current theme", () => {
        const defaultTheme = {
            colors: {
                primary: "black",
            },
        };
        let theme = null;
        let TestComponent = () => {
            theme = useTheme();
            return null;
        };
        renderer.create((
            <ThemeProvider theme={defaultTheme}>
                <TestComponent />
            </ThemeProvider>
        ));

        expect(theme).not.toBeNull();
        expect(theme?.colors?.primary).toBe(defaultTheme.colors.primary);
    });
});

describe("styled", () => {
    it("should return a valid React component", () => {
        const StyledComponent = styled("div", {
            color: "white",
        });
        const component = renderer.create((
            <StyledComponent>Hello world</StyledComponent>
        ));

        expect(component.toJSON()).toMatchSnapshot();
    });

    it("should apply theme", () => {
        const theme = {
            colors: {
                primary: "black",
            },
        };
        const StyledComponent = styled("div", {
            color: "primary",
        });
        const component = renderer.create((
            <ThemeProvider theme={theme}>
                <StyledComponent>Hello</StyledComponent>
            </ThemeProvider>
        ));
        const div = component.root.findByType("div");

        expect(div).toBeDefined();
        expect(div.props.className).toEqual(expect.stringContaining("sii-"));
        expect(document.querySelector(selector).innerHTML).toEqual(expect.stringContaining("color:black"));
    });
});
