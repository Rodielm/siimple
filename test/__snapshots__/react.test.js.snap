// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeProvider should render 1`] = `
<div>
  Hello world
</div>
`;

exports[`styled should return a valid React component 1`] = `
<div
  className="sii-311768968"
>
  Hello world
</div>
`;
